### PR TITLE
fix: ingress secret is deleted by vcluster

### LINKED ISSUE #1296: Ingress secret is deleted by vCluster when ingress sync is enabled
### What happened?

Some users are reporting that when ingress sync is enabled and a tls  enabled ingress is created, the ingress secretName is rewritten  (to avoid conflicts while mapping multiple namespaces in the vcluster to single namespaces in the host cluster). 

An secret of this name is created in the host cluster by cert-manager or other entity, but this secret is eventually deleted by the syncer. 

### What did you expect to happen?

The syncer should not delete objects it does not manage

### How can we reproduce it (as minimally and precisely as possible)?

- start a vcluster with ingress sync enabled
- install cert-manager, provision an issuer
- create an ingress in the vcluster with tls enabled
- wait for the secret to be created for the certificate in the host cluster
- restart the vCluster pod to trigger sync

I'm not sure if this is all it takes to repro, will update as I know more

### Anything else we need to know?

_No response_

### Host cluster Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Host cluster Kubernetes distribution

<details>

```
# Write here
```

</details>

### vlcuster version

<details>

```console
$ vcluster --version
# paste output here
```

</details>

### Vcluster Kubernetes distribution(k3s(default)), k8s, k0s)

<details>

```
# Write here
```

</details>

### OS and Arch

<details>

```
OS: 
Arch:
```
</details>

### PR DESCRIPTION
**Pull request by rohantmp**
Signed-off-by: Rohan CJ <rohantmp@gmail.com>

**What issue type does this pull request address?** (keep at least one, remove the others) 
/kind bugfix

**What does this pull request do? Which issues does it resolve?** (use `resolves #<issue_number>` if possible) 
resolves #1296 
resolves ENG-2226

**Please provide a short message that should be published in the vcluster release notes**
Fixed an issue where vcluster deletes an ingress secret.

**What else do we need to know?**

### CODE REVIEW

**Comment by rohantmp** [comment]
Oops, left a debug statement in there

**Comment by rohantmp** [comment]
Also need to check how the object gets enqueued at all

**Comment by FabianKramm** [comment]
@rohantmp I think the approach is not bad, why haven't we done this before? Is there any issues with this approach you can see?

**Comment by rohantmp** [comment]
Hmm, so the only thing I think might need to be handled is when an unmanaged physical object  exists and a virtual object wants to be synced by that name.

**Comment by rohantmp** [comment]
In that case, should we just throw a syncer error and call it a naming conflict?

**Comment by FabianKramm** [comment]
@rohantmp good question, yes I think returning an error as well as a emitting a syncer error is a good here.

File: pkg/controllers/syncer/syncer.go [hunk 1]
```diff
@@ -105,14 +106,23 @@ func (r *syncerController) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
-	// this is to distinguish generic and plugin syncers with the core syncers
-	if pObj != nil && r.excludePhysical(pObj) {
-		return ctrl.Result{}, nil
+	if pObj != nil {
+		// check if physical object is actually managed by vcluster or just coincidentally named
+		isManaged, err := r.syncer.IsManaged(ctx, pObj)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to check if physical object is managed: %w", err)
+		}
+		if !isManaged {
```
Comment (FabianKramm) [inline pkg/controllers/syncer/syncer.go:119]: This should be an `else if !isManaged {`

File: pkg/controllers/syncer/syncer.go [hunk 2]
```diff
@@ -105,14 +106,23 @@ func (r *syncerController) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
+			return ctrl.Result{}, nil
+		}
+		// this is to distinguish generic and plugin syncers with the core syncers
+		if r.excludePhysical(pObj) {
```
Comment (FabianKramm) [inline pkg/controllers/syncer/syncer.go:123]: I think we should move this check up and do it before IsManaged to preserve the old behaviour as much as possible

**Comment by rohantmp** [comment]
Verified the discussed event is returned when there is a conflict: 
image

File: pkg/controllers/syncer/syncer.go (hunk 2 above)
Comment (rohantmp) [reply pkg/controllers/syncer/syncer.go]: done

**Comment by rohantmp** [comment]
New message:
image

File: pkg/controllers/syncer/syncer.go (hunk 1 above)
Comment (FabianKramm) [reply pkg/controllers/syncer/syncer.go:127]: @rohantmp yes but this is semantically still part of the above statement and looks cleaner

Comment (FabianKramm) [reply pkg/controllers/syncer/syncer.go:127]: @rohantmp any updates here?

**Comment by PavelGloba** [comment]
It would be super cool if this was included in a 16.4 🙏

**Review by FabianKramm — APPROVED** [review]

**Comment by rohantmp** [comment]
@PavelGloba 
It will be!

**Comment by dimq** [comment]
Nice, thank you very much all of you for your work, we've just seen the same problem here !!! Wonderful work !!!!

**Pull request merged**